server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java

```diff
@@ -475,13 +490,50 @@ long globalOrdToBucketOrd(long owningBucketOrd, long globalOrd) {
         @Override
         void forEach(long owningBucketOrd, BucketInfoConsumer consumer) throws IOException {
             assert owningBucketOrd == 0;
-            for (long globalOrd = 0; globalOrd < valueCount; globalOrd++) {
-                if (false == acceptedGlobalOrdinals.test(globalOrd)) {
-                    continue;
+            if (excludeDeletedDocs) {
+                forEachExcludeDeletedDocs(consumer);
+            } else {
+                for (long globalOrd = 0; globalOrd < valueCount; globalOrd++) {
+                    if (false == acceptedGlobalOrdinals.test(globalOrd)) {
+                        continue;
+                    }
+                    long docCount = bucketDocCount(globalOrd);
+                    if (bucketCountThresholds.getMinDocCount() == 0 || docCount > 0) {
+                        consumer.accept(globalOrd, globalOrd, docCount);
+                    }
                 }
-                long docCount = bucketDocCount(globalOrd);
-                if (bucketCountThresholds.getMinDocCount() == 0 || docCount > 0) {
-                    consumer.accept(globalOrd, globalOrd, docCount);
+            }
+        }
+
+        /**
+         * Excludes deleted docs in the results by cross-checking with liveDocs.
+         */
+        private void forEachExcludeDeletedDocs(BucketInfoConsumer consumer) throws IOException {
+            try (LongHash accepted = new LongHash(20, NON_RECYCLING_INSTANCE)) {
```

Comment:
We should pass in the normal `BigArrays`. These `NON_RECYCLING_INSTANCE` is fairly scary and we would love to move it to test-land.